### DISCUSSION TITLE
User-defined position for search results (Pinning documents)

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#323] [Feedback & Feature Proposal]
Following a discussion initiated here https://github.com/meilisearch/product/issues/43#issuecomment-982801719 by @francoism90.

---

I'm using something like this:
```php
$client->index('movies')->search('*', ['filter' => 'id = 20, id = 10, id = 30']);
```
When no sort is giving, I expect item 20 first, 10 second and 30 third. However, the results still seems to be based on relevance.

I'm I missing something here? I cannot do `sort => id:asc` because this would result in: 10, 20, 30 ..

Thanks.

---

In this particular case, it seems to me that MeiliSearch returns the default order of the documents in the database. I am not an expert on Laravel Scout but it seems impossible to me in the current state unless a sorting on the client-side is done, I doubt it.

The sort can only be ascending or descending and is triggered after or before the tie-breaking of other ranking rules, it depends on the position of the sort ranking rule among them. For a placeholder search (aka without any term) entered by the end-user, MeiliSearch will not be able to define an order other than strict ascending or descending since no other ranking rules will be triggered to mitigate the position by relevancy.

It reminds me strongly of being able to pin results in a precise order no matter what the rankings rules return. It would be a kind of bypass. It would also be very useful in an e-commerce use case.

**Comment by francoism90** [comment]
@gmourier Thanks for creating a discussion. :)

> It reminds me strongly of being able to pin results in a precise order no matter what the rankings rules return. It would be a kind of bypass. It would also be very useful in an e-commerce use case.

Yep. It's really helpful on cases that aren't directly indexed, like user feeds (e.g. favorites, history, etc.), showing items at random or indeed 'pinning' results.

This is my workaround for now:

- When no `filter` is giving by the user: get result directly from database (e.g. user favorites) with the wanted order.
- When filter is giving by the user: get matching IDs from DB, inject them as filter `id` for MeiliSearch, return results.

You may ask why not directly use DB queries (incl. filtering)? Well, they are slow, inflexible (not every attribute is listed in the DB), lack features (`fulltext` is terrible) and hit your overall application performance. By running a separate search engine, one can prevent certain hits on the database and even can apply caching more easily.

Another workaround may be creating multiple indexes, like `user_favorites`, `user_history`, `book_popular`, `book_trending`, etc. and use that order (e.g. `id:asc`), however this would require multiple indexes and I kinda dislike the idea of adding user related data to search solutions/different storage. It's possible, but can be really difficult to maintain. What are your opinions about this solution @gmourier?

**Reply by gmourier** [reply]
Hi @francoism90,  what are your clauses to get the desired order from your main DB? If you use a sort criterion in SQL (I assume you use a SQL engine 🤓)  on a specific field to get this order, you just have to index the same field and redo the same kind of sort on it.

Are they ids relating to the same table? I have the impression from reading you that it comes from several tables `user_favorites`, `user_history` for example. Tell me if I'm wrong!

What does this `id` actually represent?

> Another workaround may be creating multiple indexes, like user_favorites, user_history, book_popular, book_trending, etc. and use that order (e.g. id:asc), however this would require multiple indexes and I kinda dislike the idea of adding user related data to search solutions/different storage. It's possible, but can be really difficult to maintain. What are your opinions about this solution @gmourier?

It's a difficult solution to maintain as long as we don't have a multi-index search but unfortunately, I don't see any other solution right now.

Feel free to contact me on our [slack community](https://slack.meilisearch.com/) to discuss it directly, it will probably be easier. I feel like there are several factors at play.

**Comment by macraig** [comment]
Hey there 👋 

We published a workaround guide for [Displaying promoted search results on our blog](https://blog.meilisearch.com/promoted-search-results-with-react-instantsearch/?utm_campaign=oss&utm_source=github&utm_medium=discussion). We hope it's useful, feedback is welcome! 

**Note:** The tutorial uses React, we're thinking about writing another tutorial explaining how to implement this workaround on the server side. Let us know if you're interested!

______________________

**August 10th update:** We added a second tutorial to implement on the server side,[How to implement document boosting with a second index](https://blog.meilisearch.com/document-boosting/)
